### DISCUSSION TITLE
Hide API Keys with PyInstaller

### DISCUSSION THREAD

**Discussion by devcodeman** [pyinstaller/pyinstaller#7520] [Help]
Hello, I am wondering how to bundle my .env for api keys securely when I utilize pyinstaller.

I have noticed that when run the following command:

pyinstaller main.py --onefile -n testbot --add-data '.env;.'

My .env file will be exposed in the windows TEMP directory. 

Any pointers on how to handle this?

**Comment by bwoodsend** [comment]
In order to use the api keys, your application needs to contain everything needed to obfuscate them so there is no such thing as a *secure* way to put any kind of secrets in any kind of application.

If you're content with merely making it awkward rather than impossible to extract the api keys, I'd suggest moving your secrets into a `.py` file then use [pyarmor](http://pyarmor.dashingsoft.com/) on it.

**Reply by rtrad89** [reply]
Are you suggesting we need the application to decrypt these API keys using a private key stored within the application? or what would you recommend?


Besides, I am on Linux (Ubuntu server), and I am wondering where that file is cached? I find that I still need to store `.env` side-by-side with the entry script for it to work, so thought the file isn't actually copied by pyinstaller.. please advise and thanks

**Reply by bwoodsend** [reply]
I'm saying that you shouldn't put API keys in the application encrypted or otherwise. A separate `.env` file is as good as it gets. 

> I am wondering where that file is cached?

What makes you think it's being cached?

**Reply by rtrad89** [reply]
> I'm saying that you shouldn't put API keys in the application encrypted or otherwise. A separate `.env` file is as good as it gets.
> 
> > I am wondering where that file is cached?
> 
> What makes you think it's being cached?

I think my confusion comes from `--add-data` and OP's comment "My .env file will be exposed in the windows TEMP directory." I thought `--add-data` adds the needed data to the bundled application. However, whether I use `--add-data` or not, I still need to copy `.env` to the same directory from which I run the application, not sure why:
```pyinstaller --onefile --add-data="/home/username/.secrets/.env:." ../Solutions/application.py```

In my application, I am loading this file via the relative path `.env`. I think this is the problem (perhaps [this](https://github.com/pyinstaller/pyinstaller/issues/5522#issuecomment-770858489) is how it should be done).

**Reply by rokm** [reply]
> In my application, I am loading this file via the relative path `.env`. I think this is the problem.

It is. 

Using relative paths is not portable and will usually not work with PyInstaller-frozen application. It will also not work if your run your `application.py` unfrozen from an arbitrary location (e.g., from /home/username).
